devdoc/Metrics.rst

```diff
@@ -0,0 +1,354 @@
+  the above.
+
+Metric Value
+  The current value of a metric. A corollary is a metric measurement which represents a view of a metric
+  value that may be based on state.
+
+Collection
+  The process of choosing the relevant metrics and retrieving their values.
+
+Trigger
+  What causes the collection of metric values.
+
+Metric Report
+  Formatted metric values ready for ingestion by a collection system.
+
+
+
+*************
+Use Scenarios
+*************
+This section describes how components expect to use the framework. It is not a complete list of all
+requirements but rather a sample.
+
+
+Roxie
+=====
+Roxie desires to keep a count of many different internal values. Some examples are
+
+* Disk type operations such as seeks and reads
+* Execution totals
+
+  Need to track items such as total numbers of items such as success and failures
+  as well as breaking some counts into individual reasons. For example, failures
+  may need be categorized such as as
+
+  * Busy
+  * Timeout
+  * Bad input
+
+  Or even by priority (high, low, sla)
+
+* Current operational levels such as the length of internal queues
+* The latency of operations such as queue results, agent responses, and gateway responses
+
+Roxie also has the need to track internal memory usage beyond the pod/system level capabilities.
+Tracking the state of its large fixed memory pool is necessary.
+
+The Roxie buddy system also must track how often and who is completing requests. The "I Beat You To It"
+set of metrics must be collected and exposed in order to detect pending node failure. While specific
+action on these counts is not known up front, it appears clear that these values are useful and should
+be collected.
+
+There does not appear to be a need for creating and destroying metrics dynamically. The set of metrics
+is most likely to be created at startup and remain active through the life of the Roxie. If, however,
+stats collection seeps into the metrics framework, dynamic creation and destruction of stats metrics is
+a likely requirement.
+
+
+ESP
+===
+
+There are some interesting decisions with respect to ESP and collection of metrics. Different
+applications within ESP present different use cases for collection. Ownership of a given task drives
+some of these use cases. Take workunit queues. If ownership of the task, with respect to metrics, is
+WsWorkunits, then use cases are centric to that component. However, if agents listening on the queue
+are to report metrics, then a different set of use cases emerge. It is clear that additional work is
+needed to generate clear ownership of metrics gathered by ESP and/or the tasks it performs.
+
+ESP needs to report the *activeTransactions* value from the TxSummary class(es). This gives an
+indication of how busy the ESP is in terms of client requests.
+
+Direct measurement of response time in requests may not be useful since the type of request causes
```

Comment:
The notes here are based on a response from Tony. I wanted to capture the thoughts.

Reply:
Right, I recall that comment, and I agree "it may not be useful" because the individual time segments that comprise the entire response time will pin-point potentials issues. But at the same time, its importance in monitoring shouldn't be underestimated, specially when all individual time segments aren't available.